# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/ChangeStyleTransaction.cpp
# Commit: ce2087ddab5c
# Full Hash: ce2087ddab5c2efcbce3b52da02b3363a37d6813
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2020-04-04 09:29:13
# Regressor Bug: 1619914
# File Overlap Count: 1
# Description:
#   Bug 1619914 - part 1: Make each transaction class grab their members with local variable before touching the DOM tree r=m_kato
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D69302
# ==============================================================================

diff -r 1c8115a9a684 -r ce2087ddab5c editor/libeditor/ChangeStyleTransaction.cpp
--- a/editor/libeditor/ChangeStyleTransaction.cpp	Fri Apr 03 10:21:33 2020 +0000
+++ b/editor/libeditor/ChangeStyleTransaction.cpp	Fri Apr 03 08:30:37 2020 +0000
@@ -24,38 +24,40 @@
 
 // static
 already_AddRefed<ChangeStyleTransaction> ChangeStyleTransaction::Create(
-    Element& aElement, nsAtom& aProperty, const nsAString& aValue) {
+    nsStyledElement& aStyledElement, nsAtom& aProperty,
+    const nsAString& aValue) {
   RefPtr<ChangeStyleTransaction> transaction =
-      new ChangeStyleTransaction(aElement, aProperty, aValue, false);
+      new ChangeStyleTransaction(aStyledElement, aProperty, aValue, false);
   return transaction.forget();
 }
 
 // static
 already_AddRefed<ChangeStyleTransaction> ChangeStyleTransaction::CreateToRemove(
-    Element& aElement, nsAtom& aProperty, const nsAString& aValue) {
+    nsStyledElement& aStyledElement, nsAtom& aProperty,
+    const nsAString& aValue) {
   RefPtr<ChangeStyleTransaction> transaction =
-      new ChangeStyleTransaction(aElement, aProperty, aValue, true);
+      new ChangeStyleTransaction(aStyledElement, aProperty, aValue, true);
   return transaction.forget();
 }
 
-ChangeStyleTransaction::ChangeStyleTransaction(Element& aElement,
+ChangeStyleTransaction::ChangeStyleTransaction(nsStyledElement& aStyledElement,
                                                nsAtom& aProperty,
                                                const nsAString& aValue,
                                                bool aRemove)
     : EditTransactionBase(),
-      mElement(&aElement),
+      mStyledElement(&aStyledElement),
       mProperty(&aProperty),
       mValue(aValue),
-      mRemoveProperty(aRemove),
       mUndoValue(),
       mRedoValue(),
+      mRemoveProperty(aRemove),
       mUndoAttributeWasSet(false),
       mRedoAttributeWasSet(false) {}
 
 #define kNullCh (char16_t('\0'))
 
 NS_IMPL_CYCLE_COLLECTION_INHERITED(ChangeStyleTransaction, EditTransactionBase,
-                                   mElement)
+                                   mStyledElement)
 
 NS_INTERFACE_MAP_BEGIN_CYCLE_COLLECTION(ChangeStyleTransaction)
 NS_INTERFACE_MAP_END_INHERITING(EditTransactionBase)
@@ -141,20 +143,20 @@
 }
 
 NS_IMETHODIMP ChangeStyleTransaction::DoTransaction() {
-  // TODO: Change mElement to RefPtr<nsStyleElement>.
-  nsCOMPtr<nsStyledElement> inlineStyles = do_QueryInterface(mElement);
-  if (NS_WARN_IF(!inlineStyles)) {
-    return NS_ERROR_INVALID_ARG;
+  if (NS_WARN_IF(!mStyledElement)) {
+    return NS_ERROR_NOT_AVAILABLE;
   }
 
-  nsCOMPtr<nsICSSDeclaration> cssDecl = inlineStyles->Style();
+  OwningNonNull<nsStyledElement> styledElement = *mStyledElement;
+  nsCOMPtr<nsICSSDeclaration> cssDecl = styledElement->Style();
 
   // FIXME(bug 1606994): Using atoms forces a string copy here which is not
   // great.
   nsAutoCString propertyNameString;
   mProperty->ToUTF8String(propertyNameString);
 
-  mUndoAttributeWasSet = mElement->HasAttr(kNameSpaceID_None, nsGkAtoms::style);
+  mUndoAttributeWasSet =
+      mStyledElement->HasAttr(kNameSpaceID_None, nsGkAtoms::style);
 
   nsAutoString values;
   nsresult rv = cssDecl->GetPropertyValue(propertyNameString, values);
@@ -221,7 +223,7 @@
   uint32_t length = cssDecl->Length();
   if (!length) {
     nsresult rv =
-        mElement->UnsetAttr(kNameSpaceID_None, nsGkAtoms::style, true);
+        styledElement->UnsetAttr(kNameSpaceID_None, nsGkAtoms::style, true);
     if (NS_FAILED(rv)) {
       NS_WARNING("Element::UnsetAttr(nsGkAtoms::style) failed");
       return rv;
@@ -238,16 +240,18 @@
 
 nsresult ChangeStyleTransaction::SetStyle(bool aAttributeWasSet,
                                           nsAString& aValue) {
+  if (NS_WARN_IF(!mStyledElement)) {
+    return NS_ERROR_NOT_AVAILABLE;
+  }
+
   if (aAttributeWasSet) {
+    OwningNonNull<nsStyledElement> styledElement = *mStyledElement;
+
     // The style attribute was not empty, let's recreate the declaration
     nsAutoCString propertyNameString;
     mProperty->ToUTF8String(propertyNameString);
 
-    nsCOMPtr<nsStyledElement> inlineStyles = do_QueryInterface(mElement);
-    if (NS_WARN_IF(!inlineStyles)) {
-      return NS_ERROR_INVALID_ARG;
-    }
-    nsCOMPtr<nsICSSDeclaration> cssDecl = inlineStyles->Style();
+    nsCOMPtr<nsICSSDeclaration> cssDecl = styledElement->Style();
 
     ErrorResult error;
     if (aValue.IsEmpty()) {
@@ -268,7 +272,10 @@
                          "nsICSSDeclaration::SetProperty() failed");
     return error.StealNSResult();
   }
-  nsresult rv = mElement->UnsetAttr(kNameSpaceID_None, nsGkAtoms::style, true);
+
+  OwningNonNull<nsStyledElement> styledElement = *mStyledElement;
+  nsresult rv =
+      styledElement->UnsetAttr(kNameSpaceID_None, nsGkAtoms::style, true);
   NS_WARNING_ASSERTION(NS_SUCCEEDED(rv),
                        "Element::UnsetAttr(nsGkAtoms::style) failed");
   return rv;